hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/compactions/Compactor.java

```diff
@@ -381,34 +377,46 @@ protected final List<Path> compact(final CompactionRequestImpl request,
       } else {
         Closeables.close(scanner, true);
       }
-      if (!finished && writer != null) {
-        abortWriter();
+      if (progress != null) {
+        synchronized (this.progress) {
+          this.progress.remove(progress);
+        }
+      }
+      if (writer != null) {
```

Comment:
We do not need to return here?

Strange...

Reply:
Yes, but see the asserts below:

    assert finished : "We should have exited the method on all error paths";
    assert writer != null : "Writer should be non-null if no error";

When we drop through to the end of this method there is a finished `writer` we can commit. I will convert those assertions to throws of `IllegalStateException` so we can see if these invariants are not holding even when assertions are not enabled.

Reply:
I removed these assertions because the latest edits have changed the code flow here.